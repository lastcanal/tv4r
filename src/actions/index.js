import { batch } from 'react-redux'

import { extractPosts, filterPosts, matchRedditPath } from '../helpers'
import reddit from '../helpers/reddit'

import {
  SELECT_SUBREDDIT,
  INVALIDATE_SUBREDDIT,
  REMOVE_SUBREDDIT,
  REQUEST_POSTS,
  RECEIVE_POSTS,
  RECEIVE_POSTS_ERROR,
  SELECT_POST,
  NEXT_POST,
  PREVIOUS_POST,
  MEDIA_FALLBACK,
  REQUEST_COMMENTS,
  RECEIVE_COMMENTS,
  RECEIVE_COMMENTS_ERROR,
  TOGGLE_FULLSCREEN,
  TOGGLE_PLAY,
  TOGGLE_AUTOPLAY,
  TOGGLE_THEME_MODE,
  PLAYER_VOLUME_UP,
  PLAYER_VOLUME_DOWN,
  PLAYER_SCAN_FORWARDS,
  PLAYER_SCAN_BACKWARDS,
  PLAYER_SCAN_ACK,
  PLAYER_JUMP_TO,
  PLAYER_JUMP_ACK,
} from '../constants'

export const selectSubreddit = subreddit => ({
  type: SELECT_SUBREDDIT,
  subreddit,
})

export const invalidateSubreddit = subreddit => ({
  type: INVALIDATE_SUBREDDIT,
  subreddit,
})

export const removeSubreddit = subreddit => ({
  type: REMOVE_SUBREDDIT,
  subreddit,
})

export const requestPosts = subreddit => ({
  type: REQUEST_POSTS,
  subreddit,
})

export const receivePosts = (subreddit, json) => ({
  type: RECEIVE_POSTS,
  subreddit,
  posts: filterPosts(extractPosts(json)),
  error: null,
  receivedAt: Date.now(),
})

export const receivePostsError = (subreddit, error) => ({
  type: RECEIVE_POSTS_ERROR,
  subreddit,
  error,
})

export const selectPost = (post, index) => ({
  type: SELECT_POST,
  post,
  index,
})

export const nextPost = posts => ({
  type: NEXT_POST,
  posts,
})

export const previousPost = posts => ({
  type: PREVIOUS_POST,
  posts,
})

export const mediaFallback = () => ({
  type: MEDIA_FALLBACK,
})

const fetchPosts = subreddit => dispatch => {
  dispatch(requestPosts(subreddit))
  return reddit
    .fetchPosts(subreddit)
    .then(response => response.json())
    .then(response => dispatch(receivePosts(subreddit, response)))
    .catch(error => dispatch(receivePostsError(subreddit, error)))
}

const shouldFetchPosts = ({ postsBySubreddit }, subreddit) => {
  const posts = postsBySubreddit[subreddit]
  if (!posts) {
    return true
  }
  if (posts.isFetching) {
    return false
  }
  return posts.didInvalidate
}

export const fetchPostsIfNeeded = subreddit => (dispatch, getState) => {
  if (shouldFetchPosts(getState(), subreddit)) {
    return dispatch(fetchPosts(subreddit))
  }
}

const shouldInvalidateSubreddit = ({ router, postsBySubreddit }, subreddit) => {
  const match = matchRedditPath(router.location.pathname)
  const posts = postsBySubreddit[subreddit]
  return (
    !posts ||
    posts.items.length === 0 ||
    (match && match.subreddit === subreddit)
  )
}

export const invalidateSubredditIfNeeded = subreddit => (
  dispatch,
  getState,
) => {
  const newSubreddit = subreddit.toLowerCase()
  if (shouldInvalidateSubreddit(getState(), newSubreddit)) {
    return dispatch(invalidateSubreddit(newSubreddit))
  }
}

export const loadSubreddit = subreddit => (dispatch) => {
  batch(() => {
    dispatch(selectSubreddit(subreddit))
    dispatch(invalidateSubredditIfNeeded(subreddit))
    dispatch(fetchPostsIfNeeded(subreddit))
  })
}

const requestComments = (post, subreddit) => ({
  type: REQUEST_COMMENTS,
  post,
  subreddit,
})

const receiveComments = (post, subreddit, response) => ({
  type: RECEIVE_COMMENTS,
  post,
  subreddit,
  comments: response,
})

const receiveCommentsError = (post, subreddit, error) => ({
  type: RECEIVE_COMMENTS_ERROR,
  post,
  subreddit,
  error,
})

const fetchComments = (post, selectedSubreddit) => dispatch => {
  dispatch(requestComments(post, selectedSubreddit))
  return reddit
    .fetchPost(post.permalink)
    .then(response => response.json())
    .catch(error =>
      dispatch(receiveCommentsError(post, selectedSubreddit, error)),
    )
    .then(response =>
      dispatch(receiveComments(post, selectedSubreddit, response)),
    )
}

export const fetchCommentsIfNeeded = () => (dispatch, getState) => {
  const { postsBySubreddit, selectedSubreddit } = getState()
  const { index } = postsBySubreddit.cursor
  const posts = postsBySubreddit[selectedSubreddit].items
  if (posts[index] && !posts[index].comments) {
    dispatch(fetchComments(posts[index], selectedSubreddit))
  }
}

export const configToggleFullscreen = () => ({
  type: TOGGLE_FULLSCREEN,
})

export const configToggleAutoplay = () => ({
  type: TOGGLE_AUTOPLAY,
})

export const configTogglePlay = () => ({
  type: TOGGLE_PLAY,
})

export const configToggleThemeMode = () => ({
  type: TOGGLE_THEME_MODE,
})

export const playerScanForwards = seconds => ({
  type: PLAYER_SCAN_FORWARDS,
  seconds,
})

export const playerScanBackwards = seconds => ({
  type: PLAYER_SCAN_BACKWARDS,
  seconds,
})

export const playerScanAck = () => ({
  type: PLAYER_SCAN_ACK,
})

export const playerJumpTo = (number) => ({
  type: PLAYER_JUMP_TO,
  percentage: (number / 10),
})

export const playerJumpAck = () => ({
  type: PLAYER_JUMP_ACK,
})

export const playerVolumeDown = () => ({
  type: PLAYER_VOLUME_DOWN,
})

export const playerVolumeUp = () => ({
  type: PLAYER_VOLUME_UP,
})

export const handleKeyboardAction = event => (dispatch, getState) => {
  const { postsBySubreddit, selectedSubreddit } = getState()
  const posts = postsBySubreddit[selectedSubreddit] &&
    postsBySubreddit[selectedSubreddit].items
  switch (event.key) {
    case '.':
    case '>':
      return dispatch(nextPost(posts))
    case 'N':
    case 'n':
      return event.shiftKey ? dispatch(nextPost(posts)) : void (0)
    case ',':
    case '<':
      return dispatch(previousPost(posts))
    case 'P':
    case 'p':
      return event.shiftKey ? dispatch(previousPost(posts)) : void (0)
    case ' ':
    case 'Enter':
    case 'k':
      return dispatch(configTogglePlay())
    case 'a':
      return dispatch(configToggleAutoplay())
    case 'f':
      return dispatch(configToggleFullscreen())
    case 'ArrowLeft':
      return dispatch(playerScanBackwards(5))
    case 'ArrowRight':
      return dispatch(playerScanForwards(5))
    case 'ArrowUp':
      return dispatch(playerVolumeUp())
    case 'ArrowDown':
      return dispatch(playerVolumeDown())
    default:
      const jumpTo = parseInt(event.key)
      if (jumpTo) {
        return dispatch(playerJumpTo(jumpTo))
      } else {
        return
      }
  }
}
