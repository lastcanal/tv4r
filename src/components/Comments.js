import React, { useEffect, useState } from 'react'
import { connect } from 'react-redux'
import Box from '@material-ui/core/Box'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import LinearProgress from '@material-ui/core/LinearProgress'
import { mediaSelector } from '../selectors'

import {
  fetchCommentsIfNeeded,
} from '../actions'

import CommentTree from './CommentTree'

const styles = ({ spacing, palette, shape }) => ({
  root: {
    borderRadius: shape.borderRadius,
  },
  comments: {
    borderRadius: shape.borderRadius,
    minHeight: ({ playerHeight }) => (playerHeight),
  },
  loading: {
    borderRadius: shape.borderRadius,
    backgroundColor: palette.background.default,
    minHeight: ({ playerHeight }) => (playerHeight),
  },
  menuButton: {
    marginRight: spacing(2),
  },
  title: {
    color: palette.text.contrastText,
  },
  spacerBottom: {
    minHeight: ({ menuHeight }) => (menuHeight + spacing(1)),
    paddingBottom: 1,
  },
  loadComments: {
    display: 'flex',
    width: '100%',
    '& a': {
      flex: 1,
      textAlign: 'center',
      color: palette.text.primary,
      textDecoration: 'none',
      cursor: 'pointer',
    },
  },
})

const Comments = ({
  comments,
  isFetchingComments,
  post,
  playerHeight,
  dispatch,
  classes,
}) => {
  const [visible, setVisible] = useState(false)
  const [fallback, setFallback] = useState(false)

  useEffect(() => {
    if (visible) {
      dispatch(fetchCommentsIfNeeded())
    }
  }, [post, visible])

  useEffect(() => {
    if (typeof IntersectionObserver === 'undefined') {
      // IntersectionObserver not supported (ie, node),
      // default to always visible and always loading
      // https://caniuse.com/#feat=intersectionobserver
      setFallback(true)
      return
    } else {
      /* istanbul ignore next */
      const rootMargin = playerHeight
      const target = document.getElementById('scroll_beacon')
      const options = {
        rootMargin: `${rootMargin}px`,
        threshold: 1,
      }
      const onIntersection = elements => {
        if (elements[0] && elements[0].isIntersecting) {
          setVisible(true)
        } else {
          setVisible(false)
        }
      }

      const observer = new IntersectionObserver(onIntersection, options)
      if (target) {
        observer.observe(target)
        return () => observer.unobserve(target)
      }
    }
  }, [playerHeight])

  const LoadComments = () => {
    return fallback && <div className={classes.loadComments}>
      <a onClick={() => setVisible(!visible)}>
        <h2>{visible ? 'Stop Loading' : 'Load'} Comments</h2>
      </a>
    </div>
  }

  if (!comments || isFetchingComments) {
    return (
      <div>
        <div className={classes.spacer}>
          <div id="comments" className={classes.loading}>
            <Box className={classes.selfPostBody} boxShadow={5}>
              {isFetchingComments
                ? <LinearProgress />
                : <LoadComments />
              }
            </Box>
          </div>
        </div>
        <div className={classes.spacerBottom}></div>
      </div>
    )
  }

  return (
    <div>
      <div className={classes.comments}>
        <div
          id="comments"
          className={`${classes.comment_container} ${classes.root}`}
        >
          {fallback && <LoadComments />}
          {comments.root.map((comment, index) => (
            <CommentTree comments={comment} key={index} />
          ))}
        </div>
      </div>
      <div className={classes.spacerBottom}></div>
    </div>
  )
}

Comments.propTypes = {
  dispatch: PropTypes.func,
  classes: PropTypes.object,
  playerHeight: PropTypes.number,
  menuHeight: PropTypes.number,
  isFetchingComments: PropTypes.bool,
  comments: PropTypes.object,
  post: PropTypes.object,
}

const mapStateToProps = ({ postsBySubreddit, selectedSubreddit, config }) => {
  const { showVideos, showImages, showNSFW } = config
  const { cursor } = postsBySubreddit
  const subreddit = postsBySubreddit[selectedSubreddit]
  const {
    comments,
    isFetchingComments,
    scope,
  } = subreddit || {
    comments: {},
  }

  const posts = subreddit?.[scope] || []
  const items = mediaSelector({ posts, showVideos, showImages, showNSFW })
  const post = items[cursor.index]
  const commentsForPost = comments?.[post?.id]

  return {
    post,
    isFetchingComments,
    comments: commentsForPost,
  }
}

export default Comments |>
  withStyles(styles) |>
  connect(mapStateToProps)
