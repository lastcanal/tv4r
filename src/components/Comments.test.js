import { Provider } from 'react-redux'

import Comments from './Comments'
import * as actions from '../actions'


describe('Comments', () => {
  it('should render empty Comments', () => {
    const wrapper = mount(
      <Provider store={
        makeStore({
          postsBySubreddit: {
            foo: {
              items: [],
              isFetching: false,
              didInvalidate: false,
            },
            cursor: { index: 0 },
          },
          selectedSubreddit: 'foo',
        }
        )}>
        <Comments />
      </Provider>,
    )

    expect(wrapper).toMatchSnapshot()
  })


  it('should render Post', () => {
    const post = {
      id: 1,
      url: 'https://example.com/video',
      comments: [{
        kind: 'Listing',
        data: {
          children: [{
            kind: 't3',
            data: {
              title: 'foo',
            },
          }],
        },
      }, {
        kind: 'Listing',
        data: {
          children: [{
            kind: 't1',
            data: {
              body: 'bar',
              replies: {
                kind: 'Listing',
                data: {
                  children: [{
                    kind: 't2',
                    data: {
                      body: 'baz',
                    },
                  }, {
                    kind: 'more',
                    data: 'AAAAAAA',
                  }],
                },
              },
            },
          }],
        },
      }],
    }

    const store = mockStore({
      postsBySubreddit: {
        foo: {
          items: [post],
          isFetching: false,
          didInvalidate: false,
        },
        cursor: { post, index: 0 },
      },
      selectedSubreddit: 'foo',
    })

    const wrapper = mount(
      <Provider store={store}>
        <Comments />
      </Provider>,
    )

    expect(wrapper).toMatchSnapshot()
  })
})
