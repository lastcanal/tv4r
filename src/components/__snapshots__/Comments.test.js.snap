// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments should render Post 1`] = `ReactWrapper {}`;

exports[`Comments should render empty Comments 1`] = `ReactWrapper {}`;
